docs/site/src/plugins/inject-code/utils.js

```diff
@@ -0,0 +1,114 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+// Pass in an array of lines from code source.
+// Return the array without blank lines in beg or end.
+exports.trimContent = (content) => {
+  let arr = content.split("\n");
+  let start = 0;
+  let end = arr.length;
+
+  while (start < end && arr[start].trim() === "") {
+    start++;
+  }
+
+  while (end > start && arr[end - 1].trim() === "") {
+    end--;
+  }
+
+  return arr.slice(start, end).join("\n");
+};
+
+// Fix the spurious whitespace when copying code
+// from the middle of source.
+exports.removeLeadingSpaces = (codeText, prepend = "") => {
+  const numSpaces = codeText.match(/^ */) ? codeText.match(/^ */)[0].length : 0;
+  if (numSpaces === 0) {
+    return [prepend, codeText].join("\n");
+  }
+  const lines = codeText.split("\n");
+
+  return [
+    prepend,
+    lines.map((line) => line.substring(numSpaces)).join("\n"),
+  ].join("\n");
+};
+
+// Options are added to the @inject command by appending
+// a space delimited list
+
+// Remove comments. TODO: Add other langs
+const removeComments = (text, options) => {
+  const cont = options.some((element) =>
+    element.toLowerCase().includes("nocomment"),
+  );
+  if (cont) {
+    return text.replace(/^\s*\/\/\/?.*$(?:\r\n?|\n)?/gm, "");
+  } else {
+    return text;
+  }
+};
+
+const removeTests = (text, options) => {
+  const cont = options.some((element) =>
+    element.toLowerCase().includes("notest"),
+  );
+  if (cont) {
+    return text
+      .replace(/\s*#\[test.*?\n.*?}(?!;)\n?/gs, "\n{{plugin-removed-test}}\n")
+      .replace(/\{\{plugin-removed-test\}\}\s*/gm, "");
+  } else {
+    return text;
+  }
+};
+
+exports.processOptions = (text, options) => {
+  // Replace all the //docs:: lines in code and license header
+  let processed = text
+
+    .replace(
+      /^\/\/\s*Copyright.*Mysten Labs.*\n\/\/\s*SPDX-License.*?\n?$/gim,
+      "",
+    )
+    .replace(/^\s*\/\/\s*docs::\/?.*\r?$\n?/gm, "");
+  processed = removeComments(processed, options);
+  processed = removeTests(processed, options);
+
+  return processed;
+};
+
+// When including a function, struct by name
+// Need to catch the stuff that comes before
+
+exports.capturePrepend = (match, text) => {
+  const numSpaces = match[1]
+    ? match[1].match(/^\n/)
+      ? match[1].length - 1
+      : match[1].length
+    : 0;
+  let preFun = text.substring(0, match.index - 1);
+  const lines = preFun.split("\n");
+  let pre = [];
+  for (let x = lines.length - 1; x > 0; x--) {
+    if (lines[x].trim() === "}" || lines[x].trim() === "") {
+      break;
+    } else {
+      pre.push(lines[x].substring(numSpaces));
+    }
+  }
+  return pre.reverse().join("\n");
+};
+
+// If opening and closing braces don't match
+// there's a problem
+exports.checkBracesBalance = (str) => {
```

Comment:
Yeah, it's already proved to be a pita. But the idea is to act as kind of a safety net. In the end, if it fails to return 'true', the author will just have to use the doc comment approach instead. I'll add that note to the documentation.

Reply:
I'm curious -- how often has it helped vs. hurt? (In how many cases has it legitimately prevented an issue?) If it's not many, then it may also be worth dropping the check as well...

Reply:
Good question. I think the original reason was because it would grab the module closing brace if the reference was for the last item and I couldn't get a proper regex to filter it out in those cases. Then I thought it might be a good, basic check. But I think there's probably a better approach for that one case and just use the fallback for any corners that might come up.

Reply:
This remains but only as a function of processing the last thing in a block/module. Not used as a check.